### DISCUSSION TITLE
Cli does not recognize primary key for postgresql

### DISCUSSION THREAD

**Discussion by debasish-raychawdhuri** [diesel-rs/diesel#3626] [Q&A] [answered]
Hi,

I am trying to run diesel migrations on postgres, but it complains that the table does not have a primary key. This is not true. Please see the code here - https://github.com/debasish-raychawdhuri/todo_cli

Also, postgres version is 14.8, diesel-cli version is 2.0.1

**Comment by weiznich** [comment]
I cannot reproduce that behavior with the given example. Did you ask https://stackoverflow.com/questions/76239696/diesel-fails-to-detect-primary-key-in-postgres-even-when-present as well? In that case: The error message says: "Diesel only supports tables with primary keys. Table **test** has no primary key", which implies that your database contains a table with the name *test* that has no primary key. The definition of that table is neither included in your example here as well as over at stackoverflow.

**Comment by debasish-raychawdhuri** [comment]
The example does include the table definition in the migrations folder. I
have the typical diesel project setup. The error is the same "Table todos
has no primary key," but the Stackoverflow question is not mine. It seems
to be something not easily reproducible. If you tell me some commands to
generate logs or something, I will do that and post the output here. I also
tried adding select pg_sleep(1) after the creation of the table and
migration did take the time, but faced the same problem.

**Reply by weiznich** [reply, marked as answer]
Well I can only say the following:
* I cannot reproduce that problem even with the provided example.
* It's unclear for me if your database table definition matches that what's in the repo
* Diesel queries the database system to see whether there is a primary key or not. So if it reports that there is no primary key the database system tells that. The relevant query is here, you can use that to check what's in your database system: https://github.com/diesel-rs/diesel/blob/216c4c41cfb3448d9b0478c62410fa05a6ff2d97/diesel_cli/src/infer_schema_internals/information_schema.rs#L123-L125

**Comment by debasish-raychawdhuri** [comment]
Well, I do understand the issue now. The database did contain an additional table named 'test,' which was not defined. removing the table fixed the issue. Thanks for the help.
